flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterTableSchemaUtil.java

```diff
@@ -0,0 +1,783 @@
+import org.apache.calcite.sql.SqlDynamicParam;
+import org.apache.calcite.sql.SqlIdentifier;
+import org.apache.calcite.sql.SqlIntervalQualifier;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNodeList;
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.util.SqlVisitor;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.table.planner.calcite.FlinkTypeFactory.toLogicalType;
+import static org.apache.flink.table.types.utils.TypeConversions.fromLogicalToDataType;
+
+/** Util to convert {@link SqlAlterTableSchema} with source table to generate new {@link Schema}. */
+public class AlterTableSchemaUtil {
+
+    private final SqlParser.Config parserConfig;
+    private final SqlValidator sqlValidator;
+    private final Consumer<SqlTableConstraint> validateTableConstraint;
+    private final Function<SqlNode, String> escapeExpression;
+
+    AlterTableSchemaUtil(
+            SqlParser.Config parserConfig,
+            SqlValidator sqlValidator,
+            Function<SqlNode, String> escapeExpression,
+            Consumer<SqlTableConstraint> validateTableConstraint) {
+        this.parserConfig = parserConfig;
+        this.sqlValidator = sqlValidator;
+        this.validateTableConstraint = validateTableConstraint;
+        this.escapeExpression = escapeExpression;
+    }
+
+    public Schema convertSchema(
+            SqlAlterTableSchema alterTableSchema, ResolvedCatalogTable originalTable) {
+        UnresolvedSchemaBuilder builder =
+                new UnresolvedSchemaBuilder(
+                        originalTable,
+                        parserConfig,
+                        (FlinkTypeFactory) sqlValidator.getTypeFactory(),
+                        sqlValidator,
+                        validateTableConstraint,
+                        escapeExpression);
+        AlterSchemaStrategy strategy = computeAlterSchemaStrategy(alterTableSchema);
+        List<SqlNode> columnPositions = alterTableSchema.getColumnPositions().getList();
+        builder.addOrModifyColumns(strategy, columnPositions);
+        alterTableSchema
+                .getWatermark()
+                .ifPresent(sqlWatermark -> builder.addOrModifyWatermarks(strategy, sqlWatermark));
+        alterTableSchema
+                .getFullConstraint()
+                .ifPresent(
+                        (pk) ->
+                                builder.addOrModifyPrimaryKey(
+                                        strategy, pk, columnPositions.isEmpty()));
+        if (strategy == AlterSchemaStrategy.MODIFY) {
```

Comment:
For the `ALTER TABLE T MODIFY` operation, it is necessary to recheck the derived column/pk constraint/watermark spec at the last step.

The reason not to put this check inside the `builder#addOrModifyColumns` is that the modification might be a batch update.

E.g. the original table has the following schema
```sql
CREATE TABLE T (
  a INT NOT NULL,
  b STRING,
  c BIGINT,
  PRIMARY KEY(a, b) NOT ENFORCED
)
```
And the following operation should be considered valid.
```sql
ALTER TABLE T MODIFY (a INT METADATA VIRTUAL, CONSTRAINT ct PRIMARY KEY(b))
```